### PR TITLE
Add support for Database.load(Array[Byte]) constructor

### PR DESCRIPTION
**Pull request by olafurpg**
This PR adds better support to load a Mirror from a single fat semanticdb, for example in Scala.js. While implementing this feature, I stumbled on several issues in the current implementation which I took liberty to fix along the way

- Multpath.value is now List[AbsolutePath]
- File IO like jar in Multipath has been refactored out into Platform specific code or is guarded against a `isJVM` boolean literal to please the Scala.js dead code eliminator.
- Support to unit tests against `Mirror` in both Scala.js and JVM

### CODE REVIEW

**Comment by olafurpg** [comment]
This PR is almost ready for review. Still need to fix the JS tests, which I will try look at tomorrow along with #867 .

**Comment by olafurpg** [comment]
This PR is ready for review now @xeno-by

**Comment by olafurpg** [comment]
This PR is ready for review now :) 

The s/Sourcepath/AbsolutePath/ change is probably the most controversial. If you don't like that change, then I would like to see a test case where we need Sourcepath. Relativization of filenames becomes much simpler when using a single `sourceroot: AbsolutePath` instead of `Sourcepath`.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala [hunk 1]
```diff
@@ -44,15 +44,13 @@ package object semantic {
           }
           val (spath, scontents) = minput match {
             case mInput.File(path, charset) if charset == Charset.forName("UTF-8") =>
-              val spath = sourcepath.relativize(path.toURI)
-              spath.getOrElse(sys.error(s"bad database: can't find $path in $sourcepath")) -> ""
+             path.toRelative(sourceroot).toString -> ""
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: The indentation is a bit weird

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Entry.scala [hunk 2]
```diff
@@ -3,10 +3,13 @@ package scala.meta.internal.semantic.vfs
...
 import org.scalameta.adt._
 import scala.collection.immutable.Seq
+import scala.meta.internal.io.FileIO
 import scala.meta.io._
 import scala.meta.internal.io.InputStreamIO
+import scala.meta.internal.platform
 
-@root trait Entry {
+@root
+trait Entry {
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Entry.scala]: I'd prefer to have this and the like on a single line

File: scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala [hunk 3]
```diff
@@ -12,6 +12,8 @@ import scala.meta.internal.semantic.{vfs => v}
 import scala.meta.internal.semantic.{schema => s}
 import scala.meta.{semantic => m}
 
+import org.scalameta.logger
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala]: Let's remove this?

File: scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala [hunk 4]
```diff
@@ -36,4 +38,10 @@ object Database {
   def load(classpath: Classpath, sourcepath: Sourcepath): Database = {
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala]: Why does this compile?

File: scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala [hunk 5]
```diff
@@ -36,4 +38,10 @@ object Database {
...
     v.Database.load(classpath).toSchema.toMeta(sourcepath)
   }
+  def load(bytes: Array[Byte]): Database = {
+    val sattrs = s.Attributes.parseFrom(bytes)
+    val sdb = new s.Database(Seq(sattrs))
+    val mdb = sdb.toMeta(Sourcepath.workingDirectory)
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala]: I don't like this. How about we just crash if we encounter a slim database?

Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala]: Changed to throw an exception if trying to load a slimdb from Array[Byte]

File: scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala (hunk 4 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala]: I have removed Sourcepath.

**Comment by xeno-by** [comment]
Do we consider that #867 is fixed by this pull request?

File: scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala [hunk 6]
```diff
@@ -20,8 +20,8 @@ import scala.meta.{semantic => m}
   lazy val denotations: Map[Symbol, Denotation] = entries.flatMap(_._2.denotations).toMap
   lazy val sugars: Map[Position, String] = entries.flatMap(_._2.sugars).toMap
 
-  def save(classpath: Classpath, sourcepath: Sourcepath): Unit = {
-    this.toSchema(sourcepath).toVfs(classpath).save()
+  def save(classpath: Classpath, sourceroot: AbsolutePath): Unit = {
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala]: `classpath` here should be `AbsolutePath` instead. This fixes another ugliness in the design that I disliked. Thank you!

File: scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala [hunk 7]
```diff
@@ -34,6 +34,12 @@ import scala.meta.{semantic => m}
 
 object Database {
   def load(classpath: Classpath, sourcepath: Sourcepath): Database = {
-    v.Database.load(classpath).toSchema.toMeta(sourcepath)
+    v.Database.load(classpath).toSchema.toMeta(Some(sourcepath))
+  }
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/semantic/Database.scala]: How about we add another overload that takes `(classpath: Classpath)`, i.e operates exactly like `Array[Byte]`, but without its users having to know things like `META-INF`?

**Comment by olafurpg** [comment]
Seems v1.8 is ready for release then 🎉

**Pull request merged**